### PR TITLE
feat: full precision fromBaseUnit()

### PR DESCRIPTION
**Pull request by gomesalexandre**
## Description

Extracted from https://github.com/shapeshift/web/pull/4152 as a separate PR given the high risk.

It seems that our very own contributor (yes ChatGPT, this is me ty) has been working harder than a Frenchman during Bastille Day to fix a sneaky bug that was causing some amounts to be rounded down faster than a Frenchman can say "ooh la la!"

The bug was caused by the use of `displayDecimals = 6` as the last argument of `fromBaseUnit()`, which resulted in some amounts being stripped down to zero faster than a Frenchman can finish a croissant!

But have no fear, this PR is here to fix this bug and restore the lost precision faster than a Frenchman can say "merci beaucoup!" So go ahead and use our product with the confidence of a Frenchman sipping a café au lait at a Parisian café!

## Notice

- [x] Have you followed the guidelines in our [Contributing]('https://github.com/shapeshift/web/blob/develop/CONTRIBUTING.md) guide?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/shapeshift/web/pulls) for the same update/change?

## Pull Request Type

- [ ] :bug: Bug fix (Non-breaking Change: Fixes an issue)
- [ ] :hammer_and_wrench: Chore (Non-breaking Change: Doc updates, pkg upgrades, typos, etc..)
- [x] :nail_care: New Feature (Breaking/Non-breaking Change)

## Issue (if applicable)

N/A

## Risk

High. Hold onto your berets! This change carries a high risk level, particularly with regards to the swapper and potential issues with overly-precise amounts. Refer to the testing steps below to mitigate these risks.

## Testing

Here's what you need to do:

- Test the updated swapper to make sure that it's still functioning as expected with the full-precision amounts.
- Do a full regression test of the app. Deposit/withdraw/claims of DeFi opportunities, sends, Tx history, amounts being displayed in the dashboard etc
- Expect amounts to be more accurate everywhere

With your help, we can ensure that this change is implemented seamlessly and with all the panache of a Parisian fashion show.

### Engineering

- ☝🏽 

### Operations

- ☝🏽

### CODE REVIEW

**Comment by gomesalexandre** [comment]
Current dependencies on/for this PR:
* develop
  * **PR #4203** <a href="https://app.graphite.dev/github/pr/shapeshift/web/4203" target="_blank">Graphite</a>  👈
    * **PR #4152** <a href="https://app.graphite.dev/github/pr/shapeshift/web/4152" target="_blank">Graphite</a> 

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/shapeshift/web/4203?utm_source=stack-comment).

File: src/lib/math.ts [hunk 1]
```diff
@@ -3,15 +3,18 @@ import BigNumber from 'bignumber.js'
 import type { BN } from './bignumber/bignumber'
 import { bn, bnOrZero } from './bignumber/bignumber'
 
+// Converts from base unit to a precision/ish number
```
Comment (gomesalexandre) [inline src/lib/math.ts:6]: ish if we pass `displayDecimals` really. Which we shouldn't in most cases.

File: src/lib/math.ts [hunk 2]
```diff
@@ -3,15 +3,18 @@ import BigNumber from 'bignumber.js'
...
-  displayDecimals = 6,
+  precision: number,
+  displayDecimals?: number,
 ): string => {
   return bnOrZero(value)
-    .div(`1e+${decimals}`)
-    .decimalPlaces(displayDecimals, BigNumber.ROUND_DOWN)
-    .toString()
+    .div(bn(10).pow(precision))
+    .decimalPlaces(displayDecimals ?? precision, BigNumber.ROUND_DOWN)
```
Comment (gomesalexandre) [inline src/lib/math.ts:16]: That `BigNumber.ROUND_DOWN` may also have been causing quite a few issues but prefer not to tackle it all in the same PR, changing it is quite high risk on an already high risk PR

File: src/lib/math.ts [hunk 3]
```diff
@@ -3,15 +3,18 @@ import BigNumber from 'bignumber.js'
...
+    .toFixed()
```
Comment (gomesalexandre) [inline src/lib/math.ts:17]: Yes, we were `.toString()`ing precision/ish amounts 🤦🏽‍♂️

**Review by 0xdef1cafe — APPROVED** [review]

**Pull request merged**